text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,902 @@
+`~const Trait` bounds cannot be made conditional with `#[cfg]`. The only real useful reason to have this is to support newer Rust versions with a cfg, and allow older Rust versions to compile
+the traits, just without const support. This is surmountable with proc macros that either generate two versions or just generate a different version depending on the Rust version.
+Since it's only necessary for a transition period while a crate wants to support both pre-const-trait Rust and
+newer Rust versions, this doesn't seem too bad. With a MSRV bump the proc macro usage can be removed again.
+
+## Can't have const methods and nonconst methods on the same trait
+
+If a trait has methods that don't make sense for const contexts, but some that do, then right now it is required to split that
+trait into a nonconst trait and a const trait and "merge" them by making one of them be a super trait of the other:
+
+```rust
+const trait Foo {
+    fn foo(&self);
+}
+trait Bar: Foo {
+    fn bar(&self);
+}
+
+impl const Foo for () {
+    fn foo(&self) {}
+}
+impl Bar for () {
+    fn bar(&self) {
+        println!("writing to terminal is not possible in const eval");
+    }
+}
+```
+
+Such a split is not possible without a breaking change, so splitting may not be feasible in some cases.
+Especially since we may later offer the ability to have const and nonconst methods on the same trait, then allowing
+the traits to be merged again. That's churn we'd like to avoid.
+
+Note that it may frequently be that such a trait should have been split even without constness being part of the picture.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## use `const Trait` bounds for conditionally-const, invent new syntax for always-const
+
+It may seem tempting to use `const fn foo<T: const Trait>` to mean what in this RFC is `~const Trait`, and then add new syntax for bounds that allow using trait methods in const blocks.
+
+Examples of possible always const syntax:
+
+* `=const Trait`
+* `const const Trait` (lol)
+* `const(always) Trait` (`pub` like)
+* `const<true> Trait` (effect generic like)
+* `const! Trait`
+
+## use `Trait<const>` or `Trait<bikeshed#effect: const>` instead of `const Trait`
+
+To avoid new syntax before paths referring to traits, we could treat the constness as a generic parameter or an associated type.
+While an associated type is very close to how the implementation works, neither `effect = const` nor `effect: const` are representing the logic correctly,
+as `const` implies `~const`, but `~const` is nothing concrete, it's more like a generic parameter referring to the constness of the function.
+Fully expanded one can think of
+
+```rust
+const fn foo<T: ~const Trait + const OtherTrait>(t: T) { ... }
+```
+
+to be like
+
+```rust
+const<const C: bool> fn foo<T>(t: T)
+where
+    T: Trait + OtherTrait,
+    <T as Trait>::bikeshed#effect = const<C>,
+    <T as OtherTrait>::bikeshed#effect = const<true>,
+{
+    ...
+}
+```
```

Comment:
For the reference-level section, this seems more understandable than the `<T as Default>::k#host = Conditionally` thing above, but maybe that's just because I have already through about the "be generic over constness" formulation quite a bit.